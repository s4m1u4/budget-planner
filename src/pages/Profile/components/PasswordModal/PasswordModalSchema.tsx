import * as yup from "yup";

export const PasswordModalSchema = yup.object({
  password: yup
    .string()
    .min(6, "Password must be at least 6 characters")
    .required("Password is required"),
  repeatedPassword: yup
    .string()
    .oneOf([yup.ref("password")], "Passwords must match")
    .required("Repeat password is required"),
});
