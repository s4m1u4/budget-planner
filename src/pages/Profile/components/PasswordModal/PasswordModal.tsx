import React, { ChangeEvent, FC, useState } from "react";
import {
  ButtonComponent,
  InputComponent,
  ModalComponent,
} from "components/shared";
import { PasswordModalSchema } from "./PasswordModalSchema";
import { useFormik } from "formik";
import { IPasswordData } from "pages/Records/types";
import { Alert } from "@mui/material";

import { ButtonGroup, Title } from "./PasswordModal.styles";

interface PasswordModalProps {
  open: boolean;
  handleClose: () => void;
  setNewPassword: (passwordData: IPasswordData) => string | null;
  onSubmit: (values: IPasswordData) => void;
}

export const PasswordModal: FC<PasswordModalProps> = ({
  open,
  handleClose,
  setNewPassword,
  onSubmit,
}) => {
  const [errorMessage, setErrorMessage] = useState<string>("");

  const handleSubmit = async (values: IPasswordData) => {
    const responseError: string | null = await setNewPassword(values);
    responseError && setErrorMessage(responseError);
    !responseError && formik.resetForm();
    !responseError && handleClose();
    onSubmit(values);
  };

  const formik = useFormik({
    initialValues: {
      password: "",
      repeatedPassword: "",
    },
    validationSchema: PasswordModalSchema,
    onSubmit: handleSubmit,
  });

  const formikHandleChange = (event: ChangeEvent): void => {
    formik.handleChange(event);
    setErrorMessage("");
  };

  return (
    <ModalComponent open={open} onClose={handleClose}>
      <Title>Create a new password</Title>
      <form onSubmit={formik.handleSubmit}>
        <InputComponent
          fullWidth
          id="password"
          label="Password"
          type="password"
          name="password"
          onBlur={formik.handleBlur}
          onChange={formikHandleChange}
          value={formik.values.password || ""}
          error={(formik.touched.password && !!formik.errors.password) || false}
          helperText={(formik.touched.password && formik.errors.password) || ""}
        />
        <InputComponent
          fullWidth
          id="repeatedPassword"
          label="Repeat password"
          type="password"
          name="repeatedPassword"
          onBlur={formik.handleBlur}
          onChange={formikHandleChange}
          value={formik.values.repeatedPassword || ""}
          error={
            (formik.touched.repeatedPassword &&
              !!formik.errors.repeatedPassword) ||
            false
          }
          helperText={
            (formik.touched.repeatedPassword &&
              formik.errors.repeatedPassword) ||
            ""
          }
        />
        {errorMessage && (
          <Alert
            variant="filled"
            severity="error"
            sx={{ marginBottom: "15px" }}
          >
            {errorMessage}
          </Alert>
        )}
        <ButtonGroup>
          <ButtonComponent
            type="submit"
            color="success"
            onClick={formik.handleSubmit}
          >
            Change
          </ButtonComponent>
          <ButtonComponent type="button" color="error" onClick={handleClose}>
            Cancel
          </ButtonComponent>
        </ButtonGroup>
      </form>
    </ModalComponent>
  );
};
