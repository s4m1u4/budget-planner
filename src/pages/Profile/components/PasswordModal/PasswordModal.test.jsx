import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import { PasswordModal } from "./PasswordModal";
import userEvent from "@testing-library/user-event";

const open = true;
const onSubmit = jest.fn();
const handleClose = jest.fn();
const setNewPassword = jest.fn();

const getTitle = () => screen.getByText(/Create a new password/i);
const getInputPassword = () => screen.getByLabelText("Password");
const getInputRepeatPassword = () => screen.getByLabelText("Repeat password");
const getButtonChange = () => screen.getByRole("button", { name: /Change/i });
const getButtonCancel = () => screen.getByRole("button", { name: /Cancel/i });

const renderComponent = async () => {
  await act(
    async () =>
      await render(
        <PasswordModal
          open={open}
          onSubmit={onSubmit}
          handleClose={handleClose}
          setNewPassword={setNewPassword}
        />
      )
  );
};

describe("Password modal component", () => {
  it("make snapshot", () => {
    const { baseElement } = render(
      <PasswordModal
        open={open}
        onSubmit={onSubmit}
        handleClose={handleClose}
        setNewPassword={setNewPassword}
      />
    );

    expect(baseElement).toMatchSnapshot();
  });

  beforeEach(() => {
    render(
      <PasswordModal
        open={open}
        onSubmit={onSubmit}
        handleClose={handleClose}
        setNewPassword={setNewPassword}
      />
    );
  });

  describe("should render", () => {
    it("a title", () => {
      expect(getTitle()).toBeInTheDocument();
    });

    it("an input with name='password'", () => {
      expect(getInputPassword()).toHaveAttribute("name", "password");
    });

    it("an input with name='repeatedPassword'", () => {
      expect(getInputRepeatPassword()).toHaveAttribute(
        "name",
        "repeatedPassword"
      );
    });

    it('a button "Change', () => {
      expect(getButtonChange()).toBeInTheDocument();
    });

    it('a button "Cancel', () => {
      expect(getButtonCancel()).toBeInTheDocument();
    });

    it("error message when submitting a form with empty inputs", async () => {
      expect(screen.queryByText("Password is required")).toBeNull();
      expect(screen.queryByText("Repeat password is required")).toBeNull();
      fireEvent.click(getButtonChange());
      await renderComponent();
      expect(screen.getByText("Password is required")).toBeInTheDocument();
      expect(
        screen.getByText("Repeat password is required")
      ).toBeInTheDocument();
    });

    it("error message when entered a short password", async () => {
      expect(
        screen.queryByText("Password must be at least 6 characters")
      ).toBeNull();
      userEvent.type(getInputPassword(), "v4d1k");
      fireEvent.click(getButtonChange());
      await renderComponent();
      expect(
        screen.getByText("Password must be at least 6 characters")
      ).toBeInTheDocument();
    });

    it("error message when passwords don't match", async () => {
      expect(screen.queryByText("Passwords must match")).toBeNull();
      userEvent.type(getInputPassword(), "v4d1k");
      userEvent.type(getInputRepeatPassword(), "v4d1kU4");
      fireEvent.click(getButtonChange());
      await renderComponent();
      expect(screen.getByText("Passwords must match")).toBeInTheDocument();
    });
  });

  it("successful form submission", async () => {
    userEvent.type(getInputPassword(), "v4d1kU4");
    userEvent.type(getInputRepeatPassword(), "v4d1kU4");

    userEvent.click(getButtonChange());

    await waitFor(() =>
      expect(onSubmit).toHaveBeenCalledWith({
        password: "v4d1kU4",
        repeatedPassword: "v4d1kU4",
      })
    );

    expect(onSubmit).toHaveBeenCalledTimes(1);
    expect(getInputPassword()).toHaveValue("");
    expect(getInputRepeatPassword()).toHaveValue("");
  });
});
