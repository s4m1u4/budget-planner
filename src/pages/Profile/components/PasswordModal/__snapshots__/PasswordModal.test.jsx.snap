// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Password modal component make snapshot 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    aria-hidden="true"
    class="MuiModal-root css-79ws1d-MuiModal-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="css-5fyqun"
      tabindex="-1"
    >
      <h3
        class="css-ngqp4j"
      >
        Create a new password
      </h3>
      <form>
        <div
          class="css-acwcvw"
        >
          <label
            class="css-zkehil"
            for="password"
          >
            Password
          </label>
          <input
            class="css-1pabijg"
            id="password"
            name="password"
            type="password"
            value=""
          />
          <p
            class="css-1ghz2lr"
          />
        </div>
        <div
          class="css-acwcvw"
        >
          <label
            class="css-zkehil"
            for="repeatedPassword"
          >
            Repeat password
          </label>
          <input
            class="css-1pabijg"
            id="repeatedPassword"
            name="repeatedPassword"
            type="password"
            value=""
          />
          <p
            class="css-1ghz2lr"
          />
        </div>
        
        <div
          class="css-awol2u"
        >
          <button
            class="css-1p4hstr"
            color="success"
            type="submit"
          >
            Change
          </button>
          <button
            class="css-egajlx"
            color="error"
            type="button"
          >
            Cancel
          </button>
        </div>
      </form>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiModal-root css-79ws1d-MuiModal-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="css-5fyqun"
      tabindex="-1"
    >
      <h3
        class="css-ngqp4j"
      >
        Create a new password
      </h3>
      <form>
        <div
          class="css-acwcvw"
        >
          <label
            class="css-zkehil"
            for="password"
          >
            Password
          </label>
          <input
            class="css-1pabijg"
            id="password"
            name="password"
            type="password"
            value=""
          />
          <p
            class="css-1ghz2lr"
          />
        </div>
        <div
          class="css-acwcvw"
        >
          <label
            class="css-zkehil"
            for="repeatedPassword"
          >
            Repeat password
          </label>
          <input
            class="css-1pabijg"
            id="repeatedPassword"
            name="repeatedPassword"
            type="password"
            value=""
          />
          <p
            class="css-1ghz2lr"
          />
        </div>
        
        <div
          class="css-awol2u"
        >
          <button
            class="css-1p4hstr"
            color="success"
            type="submit"
          >
            Change
          </button>
          <button
            class="css-egajlx"
            color="error"
            type="button"
          >
            Cancel
          </button>
        </div>
      </form>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
